src/libcrun/container.c

```diff
@@ -1891,6 +1891,11 @@ container_init (void *args, char *notify_socket, int sync_socket, libcrun_error_
 
   if (entrypoint_args->exec_func)
     {
+      // Some shims and manager on kubernetes still expect process image to be changed
+      // with given executable in spec hence perform a blank execv to change process
+      // image and ignore the output since we only want to care about exec_func
+      // See: https://github.com/containers/crun/issues/798
+      execv (exec_path, (char *[]){ NULL });
```

Comment:
I don't understand why we would need this.

The execv would just fail without any effect.

Do you have any pointer in the contained code where they check for this?

Reply:
@giuseppe Sure let me try getting the exact `chunk` in `containerd` source. But when i don't perform this `exec`, `shim` reports back with `container` not ready.

Reply:
could you try if something like the following patch would work?
```
diff --git a/src/crun.c b/src/crun.c
index 640f006..108b3e1 100644
--- a/src/crun.c
+++ b/src/crun.c
@@ -310,6 +310,12 @@ main (int argc, char **argv)
   libcrun_error_t err = NULL;
   int ret, first_argument = 0;
 
+  if (getenv ("_LIBCRUN_HANDLER"))
+    {
+      sleep (60);
+      exit (EXIT_SUCCESS);
+    }
+
   argp_program_version_hook = print_version;
 #ifdef HAVE_LIBKRUN
   if (strcmp (basename (argv[0]), "krun") == 0)
diff --git a/src/libcrun/container.c b/src/libcrun/container.c
index 6f64c2c..02a295d 100644
--- a/src/libcrun/container.c
+++ b/src/libcrun/container.c
@@ -1891,6 +1891,9 @@ container_init (void *args, char *notify_socket, int sync_socket, libcrun_error_
 
   if (entrypoint_args->exec_func)
     {
+      setenv ("_LIBCRUN_HANDLER", "FOO", 1);
+           execv ("/proc/self/exe", def->process->args);
+
       ret = entrypoint_args->exec_func (entrypoint_args->container, entrypoint_args->exec_func_arg, exec_path,
                                         def->process->args);
       if (ret != 0)
```

It is just a test to make sure an `execv` is enough

Reply:
I don' think it is safe anyway, the `exec_path` could be a valid ELF or something handled by the kernel, and it could execute without the handler intercepting it.

Reply:
one issue with the `execv` to a different handler is how we pass down the .so file we have opened in the host

Reply:
@giuseppe My first attempt was `execv ("/proc/self/exe", NULL);` without the `setenv` but it failed with same issue. Let me try with your patch.

This is the exact line in `containerd` which fails https://github.com/containerd/containerd/blob/main/pkg/cri/server/container_start.go#L90 but I am also trying to find the root of `task` created by the `shim`. Somehow `shim` task is terminated immediately if `execv` from runtime is not performed.

Reply:
@giuseppe I tried your patch it fails on the same line with following error in pod state.

```console
Events:
  Type     Reason          Age               From               Message
  ----     ------          ----              ----               -------
  Normal   Scheduled       11s               default-scheduler  Successfully assigned default/wasm-hello to system
  Warning  Failed          7s                kubelet            Error: sandbox container "4e768103da15a43789e4147352ca309fb8b598cb0c3a5ee414e88dd5cb1171c3" is not running
  Normal   Pulled          5s (x2 over 10s)  kubelet            Container image "repo/wasm-test:latest" already present on machine
  Normal   Created         3s (x2 over 7s)   kubelet            Created container wasm-hello
  Normal   SandboxChanged  3s (x2 over 6s)   kubelet            Pod sandbox changed, it will be killed and re-created.
  Warning  Failed          3s                kubelet            Error: sandbox container "6522e9960077a38ce969bc2c1ca3a971e6949eaa4d20c008b56d2e189f9c5c0c" is not running
  Warning  BackOff         1s (x3 over 2s)   kubelet            Back-off restarting failed container
```